pages/deployments/oidc/oidc_with_aws.md

```diff
@@ -0,0 +1,187 @@
+
+As part of this process:
+
+1. Choose the **Custom trust policy** role type.
+
+1. Copy the following example trust policy in the following JSON code block and paste it into a code editor:
+
+    ```json
+    {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Effect": "Allow",
+                "Principal": {
+                    "Federated": "arn\:aws\:iam:\:AWS_ACCOUNT_ID\:oidc-provider/agent.buildkite.com"
+                },
+                "Action": "sts:AssumeRoleWithWebIdentity",
+                "Condition": {
+                    "StringEquals": {
+                        "agent.buildkite.com:aud": "sts.amazonaws.com"
+                    },
+                    "StringLike": {
+                        "agent.buildkite.com:sub": "organization\:ORGANIZATION_SLUG\:pipeline:PIPELINE_SLUG\:ref\:REF\:commit\:BUILD_COMMIT\:step\:STEP_KEY"
+                    },
+                    "IpAddress": {
+                        "aws:SourceIp": [
+                            "AGENT_PUBLIC_IP_ONE",
+                            "AGENT_PUBLIC_IP_TWO"
+                        ]
+                    }
+
+                }
+            }
+        ]
+    }
+    ```
+
+1. Modify the following sections of the pasted code snippet accordingly:
+    * `AWS_ACCOUNT_ID` (in `Principal`) is your AWS account ID.
+    * next item
+
+1. In the **Custom trust policy** section,
+
+1. Specify an appropriate **Role name**, for example, `compute-ssm-oidc-example`.
+
+Set the Principal to be federated via the IAM OIDC provider we just created, and add some conditions against using this IAM role for additional security using the format of the OIDC token subject provided by Buildkite.
+
+This allows us to limit use of this IAM role to named Organizations, Pipelines, and optionally branches, commits and steps. Format is
+ `organization:ORGANIZATION_SLUG:pipeline:PIPELINE_SLUG:ref:REF:commit:BUILD_COMMIT:step:STEP_KEY`
+
+You can use a wildcard `*` to replace any subjects you don’t want to set limits on.
+
+You can also add multiple token subjects as a list if you want to use the same IAM role for multiple pipelines, or if you want a variety of branches, commits and steps enabled.
+
+If you have a public IP address associated with your Buildkite Agent's, you can also set a condition on the source IP address and further restrict access to your IAM role.
+
+Update the following sections:
+
+- In the Condition on the subject, replace:
+
+    * `ORGANIZATION_SLUG` with your Buildkite Organization
+    * `PIPELINE_SLUG` with your Pipeline
+    * `REF` - this is commonly replaced with `refs/heads/main` to enforce only the main branch using the IAM role  or `refs/tags/*` for only tagged releases able to deploy, or a wildcard if we want all branches to use it.
+    * `BUILD_COMMIT` - this is commonly replaced with a wildcard `*`
+    * `STEP_KEY` - this is commonly replaced with a wildcard `*`
```

Comment:
From your example below, it looks like these two values are optional.
Are these definitely optional values and if so, what value is assumed by omitting them in the trust policy?

Reply:
They are definitely optional, but you have to end the string in a wildcard `*` if you do, e.g. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:*:ref:refs:*",
                     ]
                 }
```

Reply:
That whole string is based on the format we set when we issue the token from the agent - see our docs here https://buildkite.com/docs/agent/v3/cli-oidc 
<img width="845" alt="Screenshot 2024-04-30 at 2 54 09 PM" src="https://github.com/buildkite/docs/assets/23623393/fc322102-8a37-416c-a138-8219ac39967f">

We can match to every single aspect of that string, or just parts on it, depending on how restrictive we want to be.

Reply:
Hmm, thanks for this additional info, I might make a note of this in my next commit to these docs.